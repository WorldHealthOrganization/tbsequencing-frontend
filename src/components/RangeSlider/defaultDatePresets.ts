const currentYear = new Date().getFullYear();

const defaultDatePresets = [
  {
    label: 'Last year',
    value: [currentYear - 1, currentYear],
  },
  {
    label: 'Last 2 years',
    value: [currentYear - 2, currentYear],
  },
  {
    label: 'Last 3 years',
    value: [currentYear - 3, currentYear],
  },
  {
    label: 'Last 5 years',
    value: [currentYear - 5, currentYear],
  },
  {
    label: 'Last Decade',
    value: [currentYear - 10, currentYear],
  },
];

export {
  defaultDatePresets,
};
