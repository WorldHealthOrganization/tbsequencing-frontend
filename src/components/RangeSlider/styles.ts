import { css } from '@emotion/react';
import appColors from '../../styles/colors';
import { font, fontSizes } from '../typography/fontSizes';

const slider = {
  margin: '30px 0',
  '.MuiSlider-rail': {
    backgroundColor: appColors.neutral.grey,
    height: '4px',
  },
  '.MuiSlider-track': {
    backgroundColor: appColors.secondary.whoBlue,
    height: '4px',
  },
  '.MuiSlider-thumb': {
    backgroundColor: appColors.secondary.whoBlue,
  },
  '.MuiSlider-valueLabelOpen, .MuiSlider-valueLabel, .MuiSlider-markLabel': {
    background: 'unset',
    color: appColors.secondary.whoBlue,
    fontFamily: font.regular,
    fontSize: fontSizes.h3,
    lineHeight: '24px',
  },
  '.MuiSlider-markLabel': {
    top: '-22px',
  },
};

const presetWrapper = css({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
  flexWrap: 'wrap',
  marginBottom: '35px',
  maxHeight: 80,
});

const presetItem = css({
  color: appColors.secondary.whoBlue,
  backgroundColor: 'unset',
  cursor: 'pointer',
  border: 'none',
  textDecoration: 'underline',
  fontSize: fontSizes.h3,
  font: font.regular,
  marginRight: '35px',
  marginBottom: '8px',
});

const sliderWrapper = {
  width: 512, padding: '24px',
};

export {
  slider,
  presetItem,
  presetWrapper,
  sliderWrapper,
};
