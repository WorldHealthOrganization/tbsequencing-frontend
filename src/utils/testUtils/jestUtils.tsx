import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
