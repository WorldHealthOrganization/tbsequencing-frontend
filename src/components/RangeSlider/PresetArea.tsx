/** @jsxImportSource @emotion/react */
import React from 'react';
import { presetItem, presetWrapper } from './styles';

interface IPresetAreaProps {
  presets: { label: string, value: number[] }[],
  onPresetClick: (value: number[]) => void;
}

const PresetArea = ({ presets, onPresetClick }: IPresetAreaProps) => (
  <div css={presetWrapper}>
    {presets.map((preset) => (
      <button
        key={preset.label}
        type="button"
        css={presetItem}
        onClick={() => onPresetClick(preset.value)}
      >
        {preset.label}
      </button>
    ))}
  </div>
);

export default PresetArea;
