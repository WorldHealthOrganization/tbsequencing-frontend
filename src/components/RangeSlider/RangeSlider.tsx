/** @jsxImportSource @emotion/react */
import React, { useState } from 'react';
import { Box, Slider } from '@mui/material';
import { AppButton } from '../AppButton/AppButton';
import { slider, sliderWrapper } from './styles';
import PresetArea from './PresetArea';
import { defaultDatePresets } from './defaultDatePresets';

interface Props {
  onSubmit: (value: number[]) => void
  min?: number,
  max?: number,
  presets?: { label: string, value: number[] }[],
  defaultValue?: number[];
}

export const RangeSlider = ({
  onSubmit, min = 0, max = 100, presets, defaultValue,
}: Props) => {
  const [sliderValue, setSliderValue] = useState<number[]>(
    defaultValue || [min, max],
  );

  if (presets) {
    presets = [
      ...defaultDatePresets,
      ...presets,
    ];
  } else {
    presets = defaultDatePresets;
  }

  const marks = [{
    value: min,
    label: min,
  }, {
    value: max,
    label: max,
  }];

  const handleSliderChange = (e: Event, value: number[]) => {
    setSliderValue(value);
  };

  const onClick = () => {
    onSubmit(sliderValue);
  };

  const onPresetClick = (value: number[]) => {
    setSliderValue(value);
  };

  return (
    <Box sx={sliderWrapper}>
      <Slider
        sx={slider}
        disableSwap
        // @ts-ignore
        onChange={handleSliderChange}
        value={sliderValue}
        marks={marks}
        min={min}
        max={max}
        valueLabelDisplay="auto"
      />
      <PresetArea presets={presets} onPresetClick={onPresetClick} />
      <AppButton variant="contained" size="medium" onClick={onClick}>
        Submit
      </AppButton>
    </Box>
  );
};

export default RangeSlider;
