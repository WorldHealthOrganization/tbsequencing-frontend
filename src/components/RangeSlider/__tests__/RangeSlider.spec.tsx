import React from 'react';
import { fireEvent, screen } from '@testing-library/react';
import { getTreeJSONWithRedux, render } from '../../../utils/testUtils/jestUtils';
import RangeSlider from '../RangeSlider';

const currentYear = new Date().getFullYear();

describe('RangeSlider component test', () => {
  test('renders RangeSlider component without error', () => {
    render(
      <RangeSlider onSubmit={() => null} />,
    );
  });

  test('calls submit func', () => {
    const submitFunc = jest.fn();
    render(<RangeSlider onSubmit={submitFunc} />);
    const button = screen.queryByText('Submit');

    if (button) {
      fireEvent.click(button);
    }

    expect(button).toBeInTheDocument();
    expect(submitFunc).toHaveBeenCalled();
  });

  test('sets value correctly after preset click', () => {
    render(<RangeSlider min={1964} max={currentYear} onSubmit={() => {}} />);

    const presetButton = screen.queryByText('Last year');

    if (presetButton) {
      fireEvent.click(presetButton);
    }

    const firstDot = screen.queryByDisplayValue(currentYear - 1);
    const secondDot = screen.queryByDisplayValue(currentYear);

    expect(presetButton).toBeInTheDocument();
    expect(firstDot).toBeInTheDocument();
    expect(secondDot).toBeInTheDocument();
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <RangeSlider onSubmit={() => null} />,
    );

    expect(tree).toMatchSnapshot();
  });
});
