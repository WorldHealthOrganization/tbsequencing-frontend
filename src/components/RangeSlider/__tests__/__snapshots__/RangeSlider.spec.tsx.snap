// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider component test Should render correctly 1`] = `
<div
  className="MuiBox-root css-1i3w57z"
>
  <span
    className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root MuiSlider-marked css-1n2xew1-MuiSlider-root"
    onMouseDown={[Function]}
  >
    <span
      className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
    />
    <span
      className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
      style={
        Object {
          "left": "0%",
          "width": "100%",
        }
      }
    />
    <span
      className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
      data-index={0}
      style={
        Object {
          "left": "0%",
        }
      }
    />
    <span
      aria-hidden={true}
      className="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
      data-index={0}
      style={
        Object {
          "left": "0%",
        }
      }
    >
      0
    </span>
    <span
      className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
      data-index={1}
      style={
        Object {
          "left": "100%",
        }
      }
    />
    <span
      aria-hidden={true}
      className="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
      data-index={1}
      style={
        Object {
          "left": "100%",
        }
      }
    >
      100
    </span>
    <span
      className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
      data-index={0}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "left": "0%",
          "pointerEvents": "none",
        }
      }
    >
      <input
        aria-orientation="horizontal"
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={0}
        data-index={0}
        disabled={false}
        max={100}
        min={0}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        step={1}
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "direction": "ltr",
            "height": "100%",
            "margin": -1,
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": "100%",
          }
        }
        type="range"
        value={0}
      />
      <span
        aria-hidden={true}
        className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
      >
        <span
          className="MuiSlider-valueLabelCircle"
        >
          <span
            className="MuiSlider-valueLabelLabel"
          >
            0
          </span>
        </span>
      </span>
    </span>
    <span
      className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
      data-index={1}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "left": "100%",
          "pointerEvents": "none",
        }
      }
    >
      <input
        aria-orientation="horizontal"
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={100}
        data-index={1}
        disabled={false}
        max={100}
        min={0}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        step={1}
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "direction": "ltr",
            "height": "100%",
            "margin": -1,
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": "100%",
          }
        }
        type="range"
        value={100}
      />
      <span
        aria-hidden={true}
        className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
      >
        <span
          className="MuiSlider-valueLabelCircle"
        >
          <span
            className="MuiSlider-valueLabelLabel"
          >
            100
          </span>
        </span>
      </span>
    </span>
  </span>
  <div
    className="css-xm0t2p-PresetArea"
  >
    <button
      className="css-1oh0tch-PresetArea"
      onClick={[Function]}
      type="button"
    >
      Last year
    </button>
    <button
      className="css-1oh0tch-PresetArea"
      onClick={[Function]}
      type="button"
    >
      Last 2 years
    </button>
    <button
      className="css-1oh0tch-PresetArea"
      onClick={[Function]}
      type="button"
    >
      Last 3 years
    </button>
    <button
      className="css-1oh0tch-PresetArea"
      onClick={[Function]}
      type="button"
    >
      Last 5 years
    </button>
    <button
      className="css-1oh0tch-PresetArea"
      onClick={[Function]}
      type="button"
    >
      Last Decade
    </button>
  </div>
  <button
    className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root css-1ve9vu2-MuiButtonBase-root-MuiButton-root-AppButton"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-3cwtjx-MuiTypography-root-BaseText"
    >
      Submit
    </p>
  </button>
</div>
`;
